!htp Fix authentication scheme of WWW-Authentication header for OAuth2 over HTTP

monkey-mas opened the pull request:
Fixes #20102

Currently, our implementation doesn't follow the [RFC 6750 section3](https://tools.ietf.org/html/rfc6750#section-3) as [reported](https://github.com/akka/akka/issues/20102#issue-142423241) by chris-zen, such that we set `Basic` scheme to WWW-Authentication header. To fix this, we need to set `Bearer` instead of `Basic` for OAuth2 authentication according to the RFC 6750 section3.

monkey-mas committed:
Fix authentication scheme of WWW-Authentication header for OAuth2 over HTTP

Currently, our implementation doesn't follow the RFC 6750 section3[1] as reported by chris-zen[2],
such that we set 'Basic' scheme to WWW-Authentication header.

To fix this, we need to set 'Bearer' instead of 'Basic' for OAuth2 authentication according to
the RFC 6750 section3.

- Not only the implementation but the corresponding test are modified within this commit.
- As challengeFor method is no longer necessary for us, the method is removed.

[1] https://tools.ietf.org/html/rfc6750#section-3
[2] https://github.com/akka/akka/issues/20102#issue-142423241

ktoso commented:
OK TO TEST

2beaucoup commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/SecurityDirectives.scala:
```diff
@@ -259,6 +259,10 @@ abstract class SecurityDirectives extends SchemeDirectives {
    * Creates a `Basic` [[HttpChallenge]] for the given realm.
    */
-  def challengeFor(realm: String): HttpChallenge = HttpChallenge.create("Basic", realm)
+  def basicChallengeFor(realm: String): HttpChallenge = HttpChallenge.create("Basic", realm)
```
`private`? Just add the scheme as a param and keep a single method?

monkey-mas replied:
If we didn't supply the method, we had the following error?
```
[info] - should define Java directive [oAuth2challengeFor] for ScalaDSL too *** FAILED *** (2 milliseconds)
[info]   akka.http.javadsl.DirectivesConsistencySpec$$anon$1: Method [akka.http.javadsl.server.directives.SecurityDirectives#oAuth2challengeFor] was not defined on class: akka.http.scaladsl.server.Directives
```
I think we need to define a Java directive as the error indicates. WDYT, @ktoso ?

ktoso replied:
Do we really need these methods here?
I think we could safely remove them. 

Perhaps adding nicer methods to `HttpChallange`
 instead.

monkey-mas replied:
I like your idea! :+1:  And yes, by applying your thought, I think we can remove these methods safely.

To make it sure, is the code below what you meant? (FYI, I do't have any preference on method names. Plz point out if you have sth in mind :) ) If the change is fine, I'm supposed to `rebase -i` and fix my first commits. 
https://github.com/monkey-mas/akka/commit/a0b4bf14771ea4c91cea86ad979898ec9ff57dec

monkey-mas replied:
Rebased and applied your idea! Can you please review them?

2beaucoup replied:
Not a fan of the extra shortcut methods.

2beaucoup commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/SecurityDirectives.scala:
```diff
@@ -254,7 +256,14 @@ trait SecurityDirectives {
    * @group security
    */
-  def challengeFor(realm: String) = HttpChallenge(scheme = "Basic", realm = realm, params = Map.empty)
+  def basicChallengeFor(realm: String) = HttpChallenge(scheme = "Basic", realm = realm, params = Map.empty)
```
private? Just add the scheme as a param and keep a single method?

monkey-mas replied:
> private?

Yeah, it seems like they should be used within `SecurityDirectives`.

> Just add the scheme as a param and keep a single method?

You mean something like this and remove the methods I've defined?
```scala
def challengeFor(scheme: String, realm: String) = HttpChallenge(scheme = scheme, realm = realm, params = Map.empty)
```

I, personally, prefer the way I did as there might be reuse of the methods, such as `def basicChallengeFor(...)`. Currently we don't reuse either of them, and your idea would be suitable :)

2beaucoup replied:
Let's keep it simple and remove these methods (scala + java). Adding tons of methods with default vals doesn't scale and they aren't reused anywhere else anyway.

monkey-mas replied:
Sure, removal done! :)

monkey-mas committed:
Add {create}Basic and {create}OAuth2 methods to HttpChallenge{s}

These methods are used to create HttpChallenge scheme "Basic" and "Bearer" respectively.
To avoid binary compatibility problem, singleton factory class in HttpChallenge.scala is
named as object HttpChallenge's'.

monkey-mas committed:
Add a new line at the end of the files `SecurityDirectives.scala` and `HttpChallenge.java`

We usually add a new line at the end of a file.

monkey-mas commented:
Oops, I think I broke binary compatibility :(
Seems like all tests passed but I've got the following log as you can see from [here](https://jenkins.akka.io:8498/job/pr-validator-per-commit-jenkins/6540/console)
```
[info] Passed: Total 849, Failed 0, Errors 0, Passed 849, Ignored 13, Pending 1
2396.423: [GC (System.gc()) [PSYoungGen: 143878K->2880K(663040K)] 1085206K->944207K(2061312K), 0.0343316 secs] [Times: user=0.18 sys=0.02, real=0.03 secs] 
2396.457: [Full GC (System.gc()) [PSYoungGen: 2880K->0K(663040K)] [ParOldGen: 941327K->695732K(1398272K)] 944207K->695732K(2061312K), [Metaspace: 385441K->380440K(1380352K)], 5.5107376 secs] [Times: user=57.37 sys=0.44, real=5.51 secs] 
2401.970: [GC (System.gc()) [PSYoungGen: 109K->384K(660992K)] 695841K->696116K(2059264K), 0.0133121 secs] [Times: user=0.11 sys=0.00, real=0.02 secs] 
2401.983: [Full GC (System.gc()) [PSYoungGen: 384K->0K(660992K)] [ParOldGen: 695732K->685932K(1398272K)] 696116K->685932K(2059264K), [Metaspace: 380440K->379791K(1380352K)], 3.8617669 secs] [Times: user=39.91 sys=0.24, real=3.86 secs] 
java.lang.RuntimeException: akka-http-core: Binary compatibility check failed!
	at scala.sys.package$.error(package.scala:27)
	at com.typesafe.tools.mima.plugin.SbtMima$$anonfun$reportErrors$1.apply(SbtMima.scala:96)
	at com.typesafe.tools.mima.plugin.SbtMima$$anonfun$reportErrors$1.apply(SbtMima.scala:78)
	at scala.collection.immutable.Map$Map4.foreach(Map.scala:181)
	at com.typesafe.tools.mima.plugin.SbtMima$.reportErrors(SbtMima.scala:78)
	at com.typesafe.tools.mima.plugin.MimaPlugin$$anonfun$mimaReportSettings$6.apply(MimaPlugin.scala:43)
	at com.typesafe.tools.mima.plugin.MimaPlugin$$anonfun$mimaReportSettings$6.apply(MimaPlugin.scala:42)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[error] (akka-http-core/*:mimaReportBinaryIssues) akka-http-core: Binary compatibility check failed!
[error] Total time: 743 s, completed Jun 10, 2016 4:56:38 PM
```

ktoso commented:
So the failure was (you can find it in the build log, or check locally via `mimaReportIssues`).

```
[error]  * the type hierarchy of object akka.http.scaladsl.model.headers.HttpChallenge is different in current version. Missing types {scala.runtime.AbstractFunction3}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("akka.http.scaladsl.model.headers.HttpChallenge$")
```

I believe we could avoid this issue by calling the object `HttpChallanges` as we do normally with those factory objects in the HTTP model. I'm honestly not quite sure if this warning can be safely ignored - let's go with the `s` suffix, as it fits the rest of the model anyway :)

monkey-mas commented:
Got it! Thanks for your advice, it's quite helpful! :)
Now that we've solved problems I suppose, please have a look at my code if you have time :D

ktoso commented:
Thanks, LGTM as far as I can see :-)
Sorry it took me a while to get back to reviewing this, ScalaDays took some of our attention the last week.

Would you be able to rebase this PR so that we can merge? Thanks in advance!

monkey-mas committed:
Remove an unnecessary scala.util.Try import from SecurityDirectives.scala

As Try isn't used, we can remove it.

monkey-mas commented:
Hi, rebased and modified the [conflict](https://github.com/monkey-mas/akka/commit/7478868b4e2ea461a747cd4c61890a61f660cfb2) into this [commit](https://github.com/akka/akka/pull/20756/commits/9ad79b1ed2a6b8a3b8b3baf4675e2d7a8db0f627)!

Can you please check it?

ktoso commented:
LGTM, thanks @monkey-mas !

monkey-mas commented:
Thanks for your help and review! :)

ktoso commented:
Thank *you* for the work! :-)
